File: 2_Procesamiento_de_datos_con_Spark/2.3_Transformaciones_y_acciones_con_DF/python/Join.py
# Databricks notebook source
# MAGIC %md
# MAGIC # Join

# COMMAND ----------

#Dataframes
data1 = [(1, "Alice"), (2, "Bob"), (3, "Charlie")]
columns1 = ["id", "name"]
df1 = spark.createDataFrame(data1, columns1)

data2 = [(1, 90), (2, 85), (4, 95)]
columns2 = ["id", "score"]
df2 = spark.createDataFrame(data2, columns2)

# Inner join
inner_join_df = df1.join(df2, on="id", how="inner")
print("inner join")
inner_join_df.show()

# Left join
left_join_df = df1.join(df2, on="id", how="left")
print("left join")
left_join_df.show()

# Right join
right_join_df = df1.join(df2, on="id", how="right")
print("right join")
right_join_df.show()

# Full outer join
full_outer_join_df = df1.join(df2, on="id", how="outer")
print("outer join")
full_outer_join_df.show()

# Cross join
cross_join_df = df1.crossJoin(df2)
print("crossjoin")
cross_join_df.show()


# COMMAND ----------

# DBTITLE 1,Join on multiple columns with matching names


df1 = spark.createDataFrame([(1, "A", 100), (2, "B", 200)], ["id", "code", "value"])
df2 = spark.createDataFrame([(1, "A", 90), (2, "C", 85)], ["id", "code", "score"])

result_df = df1.join(df2, 
                     on=["id", "code"],
                      how="inner")

result_df.show()

# COMMAND ----------

# DBTITLE 1,Join on column with different names
df1 = spark.createDataFrame([(1, "Alice"), (2, "Bob")], ["id", "name"])
df2 = spark.createDataFrame([(1, 90), (2, 85)], ["user_id", "score"])

result_df = df1.join(df2,
                     df1["id"] == df2["user_id"])

display(result_df)

File: 2_Procesamiento_de_datos_con_Spark/2.3_Transformaciones_y_acciones_con_DF/scala/Join.scala
// Databricks notebook source
import org.apache.spark.sql.functions._


// DataFrames
val data1 = Seq((1, "Alice"), (2, "Bob"), (3, "Charlie"))
val columns1 = Seq("id", "name")
val df1 = data1.toDF(columns1: _*)

val data2 = Seq((1, 90), (2, 85), (4, 95))
val columns2 = Seq("id", "score")
val df2 = data2.toDF(columns2: _*)

// Inner join
val innerJoinDF = df1.join(df2, "id")
println("Inner join")
innerJoinDF.show()

// Left join
val leftJoinDF = df1.join(df2, Seq("id"), "left")
println("Left join")
leftJoinDF.show()

// Right join
val rightJoinDF = df1.join(df2, Seq("id"), "right")
println("Right join")
rightJoinDF.show()

// Full outer join
val fullOuterJoinDF = df1.join(df2, Seq("id"), "outer")
println("Full outer join")
fullOuterJoinDF.show()

// Cross join
val crossJoinDF = df1.crossJoin(df2)
println("Cross join")
crossJoinDF.show()




// COMMAND ----------

// DBTITLE 1,Join on multiple columns with matching names

val df3 = Seq((1, "A", 100), (2, "B", 200)).toDF("id", "code", "value")
val df4 = Seq((1, "A", 90), (2, "C", 85)).toDF("id", "code", "score")

val resultDF = df3.join(df4, Seq("id", "code"), "inner")

display(resultDF)

// COMMAND ----------

// DBTITLE 1,Join on columns with different names
 val df5 = Seq((1, "Alice"), (2, "Bob")).toDF("id", "name")
val df6 = Seq((1, 90), (2, 85)).toDF("user_id", "score")

val resultDF2 = df5.join(df6, df5("id") === df6("user_id"))

display(resultDF2)
